text/006-local-js-dependencies.md

```diff
@@ -0,0 +1,280 @@
+
+* Browser-oriented without a bundler (`--browser`) intended for consumption in
+  any web browser supporting JS ES modules that also supports wasm. This mode is
+  explicitly not intended to be usable with bundlers like webpack.
+
+The `--no-modules` flag doesn't really fit any more as the `--browser` use case
+is intended to subsume that. Note that the this RFC proposes only having the
+bundler-oriented and browser-oriented modes supporting local JS snippets for
+now, while paving a way forward to eventually support local JS snippets in
+Node.js. The `--no-modules` could eventually also be supported in the same
+manner as Node.js is (once we're parsing the JS file and rewriting the exports),
+but it's proposed here to generally move away from `--no-modules` towards
+`--browser`.
+
+For some more detail about the `--browser` output, it's intended to look from
+the outside like `--no-modules` does today. When using `--browser` a single
+`wasm_bindgen` function will be exported from the module. This function takes
+either a path to the wasm file or the `WebAssembly.Module` itself, and then it
+returns a promise which resolves to a JS object that has the full wasm-bindgen
+interface on it.
+
+### JS files depending on other JS files
+
+One tricky point about this RFC is when a local JS snippet depends on other JS
+files. For example your JS might look like:
+
+```js
+// js/foo.js
+
+import { foo } from '@some/npm-package';
+import { bar } from './bar.js'
+
+// ...
+```
+
+As designed above, these imports would not work. It's intended that we
+explicitly say this is an initial limitation of this design. We won't support
+imports between JS snippets just yet, but we should eventually be able to do so.
+
+In the long run to support `--nodejs` we'll need some level of ES module parser
+for JS. Once we can parse the imports themselves it would be relatively
+straightforward for `#[wasm_bindgen]`, during expansion, to load transitively
+included files. For example in the file above we'd include `./bar.js` into the
+wasm custom section. In this future world we'd just rewrite `./bar.js` (if
+necessary) when the final output artifact is emitted. Additionally with NPM
+package support in `wasm-pack` and `wasm-bindgen` (a future goal) we could
+automatically add entries to `package.json` (or validate they're already
+present) based on the imports found.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* The initial RFC is fairly conservative. It doesn't work with `--nodejs` out of
+  the gate nor `--no-modules`. Additionally it doesn't support JS snippets
+  importing other JS initially. Note that all of these are intended to be
+  supported in the future, it's just thought that it may take more design than
+  we need at the get-go for now.
+
+* JS snippets must be written in vanilla ES module JS syntax. Common
+  preprocessors like TypeScript can't be used. It's unclear how such
+  preprocessed JS would be imported. It's hoped that JS snippets are small
```

Comment:
Maybe worth clarifying that authors of crates with local js snippets could always run the typescript compiler manually by hand. Just `wasm-bindgen` wouldn't automatically do it for you.

Reply:
this is something that wasm-pack could help with i think!